pkg/controllers/scheduling/cluster_event_handler.go

```diff
@@ -0,0 +1,94 @@
+package scheduling
+
+import (
+	clusterlisterv1alpha1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1alpha1"
+	clusterapiv1 "github.com/open-cluster-management/api/cluster/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+)
+
+type clusterEventHandler struct {
+	clusterSetLister        clusterlisterv1alpha1.ManagedClusterSetLister
+	clusterSetBindingLister clusterlisterv1alpha1.ManagedClusterSetBindingLister
+	placementLister         clusterlisterv1alpha1.PlacementLister
+	enqueuePlacementFunc    enqueuePlacementFunc
+}
+
+func (h *clusterEventHandler) OnAdd(obj interface{}) {
+	h.onChange(obj)
+}
+
+func (h *clusterEventHandler) OnUpdate(oldObj, newObj interface{}) {
+	newCluster, ok := newObj.(*clusterapiv1.ManagedCluster)
+	if !ok {
+		return
+	}
+	h.onChange(newObj)
+
+	if oldObj == nil {
+		return
+	}
+	oldCluster, ok := oldObj.(*clusterapiv1.ManagedCluster)
+	if !ok {
+		return
+	}
+
+	// if the clusterset of the cluster changes, process the original clusterset
+	if newCluster.Labels[clusterSetLabel] != oldCluster.Labels[clusterSetLabel] {
+		h.onChange(oldCluster)
+	}
+}
+
+func (h *clusterEventHandler) OnDelete(obj interface{}) {
+	h.onChange(obj)
+}
+
+func (h *clusterEventHandler) onChange(obj interface{}) {
+	// skip resource other than ManagedCluster
+	cluster, ok := obj.(*clusterapiv1.ManagedCluster)
+	if !ok {
+		return
+	}
+
+	clusterSetName, err := h.getClusterSetName(cluster)
+	if err != nil {
+		klog.Errorf("Unable to get clusterset of cluster %q: %v", cluster.Name, err)
```

Comment:
followup: this is going to create a lot of noise on startup if the clusterset informer is not synchronized.  It's not a correctness issue as we discussed in the comment below, but I think you should move this to v(4) or so.  If it's a line you need during steady state, you'll need to track the difference between synced and unsynced.

Reply:
Done. Moved the error to v(4) info